opt_adapt/opt.py

```diff
@@ -155,45 +320,60 @@ def _BFGS(it, forward_run, m, params, u, u_, dJ_, B, Rspace=False):
         P = -float(dJ) / B
         lr = line_search(forward_run, m, u, P, J, dJ, params, Rspace)
         u += lr * P
-        yield {"lr": lr, "u+": u, "u-": u_, "dJ-": dJ_, "B": B}
-        return
+        yield {"lr": lr, "u+": u, "B": B}
 
-    if B is None:
-        B = Matrix(u.function_space())
-
-    P = B.scale(-1).solve(dJ)
-    lr = line_search(forward_run, m, u, P, J, dJ, params, Rspace)
-    u += lr * P
-
-    if u_ is not None and dJ_ is not None:
-        dJ_ = params.transfer_fn(dJ_, dJ.function_space())
-        u_ = params.transfer_fn(u_, u.function_space())
-
-        s = u.copy(deepcopy=True)
-        s -= u_
-        y = dJ.copy(deepcopy=True)
-        y -= dJ_
+    else:
+        if B is None:
+            B = Matrix(u.function_space())
+        
+        # Take a step downhill
+        P = B.scale(-1).solve(dJ)
+        lr = line_search(forward_run, m, u, P, J, dJ, params, Rspace)
+        u += lr * P
+        
+        # Update the approximated inverted Hessian matrix
+        if u_ is not None and dJ_ is not None:
+            dJ_ = params.transfer_fn(dJ_, dJ.function_space())
+            u_ = params.transfer_fn(u_, u.function_space())
```

Comment:
This looks like the old version of BFGS. If not, why are you making these changes?

Reply:
Sorry, doctor, I don't remember I changed BFGS. I think it is the same as before?

Reply:
But if you look at the "Files changed" tab then there are changes within this function.

Reply:
I think I just added an else in line 325. I changed the structure to 
if Rspace 
   ------
   yield ---
else
   yield -----

my previous structure is:
if Rspace
    yield-----
    return
-------
yield------

Reply:
But I didn't change the specify functions, just changed the structure.